batch/batch/worker/worker.py

```diff
@@ -408,33 +415,39 @@ def __init__(self, job, name, spec):
         self.overlay_path = None
 
         self.image_config = None
+        self.image_digest = None
         self.rootfs_path = None
         scratch = self.spec['scratch']
         self.container_scratch = f'{scratch}/{self.name}'
         self.container_overlay_path = f'{self.container_scratch}/rootfs_overlay'
         self.config_path = f'{self.container_scratch}/config'
 
+        self.container_name = f'batch-{self.job.batch_id}-job-{self.job.job_id}-{self.name}'
+
         self.netns: Optional[NetworkNamespace] = None
         self.process = None
 
-    async def run(self, worker):
+    async def run(self, worker: 'Worker'):
         try:
 
             async def localize_rootfs():
-                async with worker.rootfs_locks[self.image_ref_str]:
+                cond = worker.rootfs_locks[self.image_ref_str]
+                async with cond:
+                    if len(worker.rootfs_digest_refs) == 10:
```

Comment:
I'm wary of us capping the number of in-use images to 10. I think we should ultimately treat this as a resource like the disks and monitor how many jobs are waiting for a root filesystem and for how long. The scheduler should probably take this number into account. Or we have a unified storage system and this won't matter anymore.

I'm still not sure why we don't use xfs copy on write here so that the root filesystem is attributed to the users quota. Then you can get rid of the underlying copy and the inodes will still exist for the in-use copy, but then you can replace the cache with another image. I think I thought about this when I was trying to design an input file cache shared between users. Take a look at `--reflink`.

Reply:
Aside from providing a more accurate account of the space left on the worker disk, I'm confused about what effect this would have. The root filesystems are readonly since they are the lowerdir of the overlay so there would be no copy on write. This feels like it would primarily introduce a large copy overhead when a worker gets fully saturated with jobs from the same batch that all want to use the same image.

Reply:
I also really don't like this arbitrary 10 business, but I would have to think for longer about a reasonable solution without a unified storage system (which feels like the actual solution).

Reply:
Now have a loop that will delete 1 unused image each minute.